### PR TITLE
save and aggregate quiz results

### PR DESCRIPTION
**Pull request by johnduffell**
In quizzes we want people to see a comparison of their results against everyone else's as they answer each question.  As such we need a way to collect the data from each user and supply an aggregate.

This endpoint will accept results at the end of the quiz and supply aggregate results at the start so that they can get the info.

note: This will be somewhere else in the long term, possibly ophan.  And we're using memcached on purpose because we're just testing the concept and don't want a datastore lying around afterwards.
Also there is no locking so it will only be approximately right as some users will overwrite each other.

Next step is to write in something to the quiz JS to use this endpoint.

@robertberry @stephanfowler what do you think?

### CODE REVIEW

File: diagnostics/app/controllers/QuizzesController.scala [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+import scala.concurrent.ExecutionContext.Implicits
+
+object QuizzesController extends Controller with Logging {
+
+  implicit val ec = Implicits.global
+
+  def results(quizId: String) = Action.async { implicit request =>
+    Quizzes.results(quizId).map {
+      json =>
+        NoCache(Ok(json))
```
Comment (kelvin-chappell) [inline diagnostics/app/controllers/QuizzesController.scala]: Does this need to be completely uncached?

File: diagnostics/app/model/diagnostics/quizzes/Quizzes.scala [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+
+  implicit val memcacheDCodec: Codec[QuizAggregate] = new Codec[QuizAggregate] {
+
+    override def serialize(value: QuizAggregate): Array[Byte] = Serialization.write(value).getBytes("UTF-8")
+
+    override def deserialize(data: Array[Byte]): QuizAggregate = parse(new String(data, "UTF-8")).extract[QuizAggregate]
+
+  }
+
+  lazy val host = Configuration.memcached.host.head
```
Comment (kelvin-chappell) [inline diagnostics/app/model/diagnostics/quizzes/Quizzes.scala]: Is `host` guaranteed to have a value?

File: diagnostics/test/quizzes/QuizzesTest.scala [hunk 3]
```diff
@@ -0,0 +1,46 @@
+package quizzes
+
+import model.diagnostics.javascript.Metric
+import model.diagnostics.quizzes.Quizzes
+import org.scalatest.{BeforeAndAfterEach, DoNotDiscover, FlatSpec, Matchers}
+
+class QuizzesTest extends FlatSpec with Matchers with BeforeAndAfterEach {
```
Comment (kelvin-chappell) [inline diagnostics/test/quizzes/QuizzesTest.scala]: Can't see any use of `BeforeAndAfterEach`

**Comment by kelvin-chappell** [comment]
If this is experimental should probably be behind a feature switch.
:+1:

File: diagnostics/app/controllers/QuizzesController.scala (hunk 1 above)
Comment (johnduffell) [inline diagnostics/app/controllers/QuizzesController.scala]: Good point, I can give it 20 mins easily enough

File: diagnostics/app/model/diagnostics/quizzes/Quizzes.scala (hunk 2 above)
Comment (johnduffell) [inline diagnostics/app/model/diagnostics/quizzes/Quizzes.scala]: looks like it - but do you think I should guard against it anyway?
see https://github.com/guardian/platform/blob/8dab73dcf47f2b0853d7f1b67c09a5a62ba1470e/provisioning/Vagrantfile#L321

Comment (kelvin-chappell) [inline diagnostics/app/model/diagnostics/quizzes/Quizzes.scala]: I was just thinking that as a general principle external developers should be able to run all our code on their machine.

**Pull request merged**